Python/avl.py

```diff
@@ -0,0 +1,276 @@
+            node.left = self._insert(node.left, value)
+        else:
+            node.right = self._insert(node.right, value)
+
+        node.height = max(self.get_height(node.left), self.get_height(node.right)) + 1
+
+        return self._update_balance(node, value)
+
+    def search(self, value):
+        '''Function to search whether the node is present in the AVL Tree or not.'''
+        if self.root is None:
+            return False
+        node = self._search(self.root, value)
+        if node:
+            logging.debug('Node found.')
+            return node
+        else:
+            logging.debug('Node not found.')
+            return False
+
+    def _search(self, node, value):
+        '''Recursive function to search the value in the AVL Tree'''
+
+        # Best case: value is present or if node does not exist
+        if node is None or node.value == value:
+            return node
+
+        # Since AVL Tree follows property of BST, if value is lesser than current node's value,
+        # it will be present in left sub tree.
+        elif value < node.value:
+            return self._search(node.left, value)
+
+        # If value is greater than current node's value then it will be present in right sub tree.
+        else:
+            return self._search(node.right, value)
+
+    def delete(self, value):
+        '''Function to delete a value from the AVL Tree.'''
+        self.root = self._delete(self.root, value)
+
+    def _delete(self, node, value):
+        '''Recursive function to delete a node from the AVL Tree.'''
+        if node is None:
+            return node
+        elif value < node.value:
+            node.left = self._delete(node.left, value)
+        elif value > node.value:
+            node.right = self._delete(node.right, value)
+        else:
+            # Node has right child present or it does not have both childs
+            if node.left is None:
+                temp = node.right
+                node = None
+                return temp
+
+            # Node has left child present.
+            elif node.right is None:
+                temp = node.left
+                node = None
+                return temp
+
+            # Node has both child present
+            # Get the in-order successor element in the right subtree, and replace
+            # it by the node to be deleted and delete the in-order successor.
+            temp = self._get_inorder_successor(node.right)
+            node.value = temp.value
+            node.right = self._delete(node.right, temp.value)
+
+        # Update the heights of the node after deletion
+        node.height = 1 + max(self.get_height(node.left), self.get_height(node.right))
+
+        # Rebalance the node
+        return self._update_balance(node, node.value)
+
+    def in_order(self, node):
+        '''Recursive In Order traversal of the tree'''
+        if node is not None:
+            self.in_order(node.left)
+            print(node.value, end=' ')
```

Comment:
printing values to stdout is often considered a bad practice. Either return a collection of node values or input a logger from the user where s/he wants the output to be written.

Reply:
Okay, will make these changes